[Google analytics] - Enable GA4 for Pegasus with unified library 1/n

pablo-code-org opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->

## Links
This PR includes the work needed:
Create a common .js file that can be used by multiple files. The common operations of:
- Creating google universal and google analytics tags
- Reading cookies
- Custom dimensions

Rewrites and simplifies the haml file (moved from erb)

Enables Pegasus to use the new library and test GA and GU properties

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story
Console
Making sure new GA4 properties are setup correctly on the datalayer
![image](https://github.com/code-dot-org/code-dot-org/assets/108825710/3ea53dd5-2a32-4a83-92eb-3faf9b1c8b5c)
Checking with Google Tag assistance debugger (http://localhost.code.org:3000/?gtm_debug=1685053948742)

The tag contains the information needed
![image](https://github.com/code-dot-org/code-dot-org/assets/108825710/0376e767-0de2-4fe0-b29a-3761f77007ad)

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+class GoogleAnalyticsReporter {
+  enableGoogleAnalyticsUniversal = true;
+  enabledGoogleAnalytics4 = true;
+  googleUniversalGtag = 'UA-37745279-1';
+  googleAnalyticsGTag = 'G-L9HT5MZ3HD';
```
ultra nit: `Gtag` or `GTag`?

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+  googleAnalyticsGTag = 'G-L9HT5MZ3HD';
+
+  constructor() {
+    this.#initializeGoogleUniversal(
```
Why are there `#`s?

cat5inthecradle replied:
Is this for ES12 Private Class Fields?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

If so, it might not be compatible with our minimum Safari and Firefox versions:

https://code.org/educate/it

pablo-code-org replied:
yes, it is for private class. Should this not be included because of safari and firefox? :(

cat5inthecradle replied:
We either need to not use it or make an argument to bump our minimum version :grin:

pablo-code-org replied:
we don't need to use it. I am just a fancy man with fancy needs. 
I can remove it and they will be public

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+    window.trackEvent = this.trackEvent;
+    window.readCookie = this.readCookie;
+  }
+  /**
```
ultra nit: add blank line.

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+ * - Centralize common code
+ */
+class GoogleAnalyticsReporter {
+  enableGoogleAnalyticsUniversal = true;
+  enabledGoogleAnalytics4 = true;
```
nit: inconsistent tense between `enable` and `enabled`.  Should be consistent.  I think maybe `isGoogleAnalytics4Enabled` makes for slightly more readable code where the value is consumed in other places.

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+  /**
+   * Enable Google Universal. This adds the tags and the ga function to the window
+   * to be used across the site.
+   */
```
This code looks like it's borrowed from somewhere.  Maybe a comment on its source?

cat5inthecradle replied:
added a suggestion for a modified comment.

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+    }
+  }
+
+  /**
+   * Common function used to read properties from cookies.
+   * Properties are extracted so they can be logged later.
+   * @name  name of the desired cookie to be extracted
+   */
```
It seems likely we already have code to do this elsewhere in our codebase that we could share?

pablo-code-org replied:
this code was so spread around already :( 

I haven't really enabled cookie reading

pablo-code-org replied:
I am sure there are more places where either I need to remove the other code OR need to export the main function here to recycle existing code. For now, I try keep it with the least disruptive version as possible (which was the existing code)

breville commented on apps/src/sites/code.org/pages/views/theme_google_analytics.js:
```diff
@@ -0,0 +1,11 @@
+import googleAnalyticsReporter from '@cdo/apps/lib/util/GoogleAnalyticsReporter';
+
+const {userAnalyticsDimensions = {}} = window;
```
Not familiar with this syntax; what does it do?

cat5inthecradle replied:
This is destructuring with a default value. Functionally the same as:

```js
const userAnalyticsDimensions = window.userAnalyticsDimensions ?? {};
```

I didn't realize you could provide default values here, but it looks handy in this case where we can't be certain that the global `window.userAnalyticsDimensions` has been initialized.

breville commented on apps/Gruntfile.js:
```diff
@@ -801,6 +801,8 @@ describe('entry tests', () => {
       './src/sites/code.org/pages/views/share_privacy.js',
     'code.org/views/theme_common_head_after':
       './src/sites/code.org/pages/views/theme_common_head_after.js',
+    'code.org/views/theme_google_analytics':
+      './src/sites/code.org/pages/views/theme_google_analytics.js',
```
Were there situations/pages where the existing JS outputs weren't enough?

pablo-code-org replied:
I don't think we had a single output containing the initialization of google Universal or setting up dimension, events or cookies before. It used to be in an ebr file 
and the other sites were doing their own initialization and code duplication. 

Also keep in mind this is the first time I am doing this^

I decided to crete its own theme so it can be reused and inject it when/where needed

breville commented on pegasus/sites.v3/code.org/views/theme_google_analytics.haml:
Pegasus does have a concept of `theme`, but is this actually a theme?

pablo-code-org replied:
it is?

I followed the convention here. I think it is

cat5inthecradle commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,152 @@
+  }
+  /**
+   * Enable Google Universal. This adds the tags and the ga function to the window
+   * to be used across the site.
```
Regarding @breville's comment about citing the source of this pasted code.

```suggestion
   * Enable Google Universal (deprecated 2023-07-01). This boilerplate code from
   * Google Analytics adds the tags and the `ga` function to the window to be used
   * across the site.
```

cat5inthecradle reviewed (changes requested):
changes requested: All of Brendan's nits that he was too nice to be a blocker on :smiling_imp:

pablo-code-org commented:
Addressed comments:
- Follow up, do more code centralization for readCookie code
- Do we need # for private methods? Should I remove it?


Feel free to reopen any convo I marked as resolved and it wasn't

pablo-code-org commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,156 @@
+    this.#addEventListener();
+
+    window.trackEvent = this.trackEvent;
+    window.readCookie = this.readCookie;
```
I haven't found where this is used besides extracting the language for hour of code which I now do using the request -_- I think it is unnecessary code but gotta check.

I am very againts removing big functions with already big changes

breville replied:
Were you able to learn any more about this?

Hamms commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,156 @@
+  /**
+   * Common function used to read properties from cookies.
+   * Properties are extracted so they can be logged later.
+   * @name  name of the desired cookie to be extracted
```
super-tiny nit: 
```suggestion
   * @param {string} name - name of the desired cookie to be extracted
```

Hamms reviewed (approved):
I'm not familiar with the details of our needs for this feature as a larger work item, but from a low level perspective just looking at the javascript and haml changes in isolation this all makes sense to me, and seems like a distinct improvement to how we organized everything previously.

breville commented:
Does this change cover https://hourofcode.com as well as https://code.org?

breville commented:
Also, does this change work on the https://code.org/ homepage in particular?  I recall it had a special set of optimizations which meant it loaded a subset of our regular pegasus pages' content.

breville reviewed (approved):
One question but otherwise looks promising!

pablo-code-org commented:
Merging, drone error is unrelated (Chrome_teacher_tools_teacher_dashboard_teacher_dashboard_eyes ) and we already had a successful run.

breville commented on apps/src/lib/util/GoogleAnalyticsReporter.js:
```diff
@@ -0,0 +1,156 @@
+  /**
+   * Public function exposed to track custom events in Google Analytics
+   */
+  trackEvent(category_value, action_name, dimensions) {
+    if (this.isGoogleAnalyticsUniversalEnabled) {
+      ga('send', 'event', category_value, action_name, dimensions);
+    }
+    if (this.isGoogleAnalytics4Enabled) {
+      gtag('event', action_name, {
+        eventCategory: category_value,
+        dimension5: 'custom data',
+      });
+    }
+  }
```
I missed this earlier, but on dashboard, `this` appears to be `Window`, rather than the singleton instance of `GoogleAnalyticsReporter`.  From a quick test, it appears that both references to `this` here should be to `googleAnalyticsReporter` instead, and then the code works as intended.
